### PR TITLE
Review arithmetic operations to be safe and sound

### LINKED ISSUE #828: Review all arithmetic operations for potential overflows
### File Location(s)

_No response_

### Proposal

All arithmetic operations should be reviewed for finding potential overflow issues.

### PR DESCRIPTION
**Pull request by shamilsan**
Resolve #828.

- Some functions have been refactored.
- Some operations have been replaced with saturating alternatives.
- Dead code was removed.

@gear-tech/dev

### CODE REVIEW

File: runtime-interface/src/lib.rs [hunk 1]
```diff
@@ -171,42 +171,37 @@ pub trait GearRI {
...
         }
 
-        // We add this page num to lazy pages in order to be able correctly
-        // finish lazy pages handling in loop. This last page won't be
-        // handled.
-        lazy_pages.push(u32::MAX);
+        let mprotect = |start, count, protect: bool| unsafe {
+            let addr = wasm_mem_addr + (start * PageNumber::size()) as u64;
+            let size = count * PageNumber::size();
+            sys_mprotect_interval(addr, size, !protect, !protect, false)
```
Comment (shamilsan) [inline runtime-interface/src/lib.rs:189]: I've just rewritten the function. @grishasobol please check whether there is no error in `protect` vs `!protect`.

File: core/src/memory.rs [hunk 2]
```diff
@@ -166,48 +166,62 @@ impl core::ops::Sub for WasmPageNumber {
...
-            pages_iter
-                .next()
-                .expect("There is wasm page, which has not all gear pages in the end");
-        }
-    }
+    let mut wasm_pages = BTreeSet::new();
+    pages_iter
+        .step_by(PageNumber::num_in_one_wasm_page() as _)
+        .try_for_each(|gp| {
+            if gp.0 % PageNumber::num_in_one_wasm_page() == 0 {
```
Comment (gshep) [inline core/src/memory.rs:192]: Sorry but I cannot understand how this condition could be true if `step_by` is called earlier?

File: core/src/memory.rs [hunk 3]
```diff
@@ -166,48 +166,62 @@ impl core::ops::Sub for WasmPageNumber {
...
-        let wasm_page_num = WasmPageNumber(page.0 / PageNumber::num_in_one_wasm_page());
-        wasm_pages.insert(wasm_page_num);
-        for _ in 0..(WasmPageNumber::size() / PageNumber::size() - 1) {
-            pages_iter
-                .next()
-                .expect("There is wasm page, which has not all gear pages in the end");
-        }
-    }
+    let mut wasm_pages = BTreeSet::new();
+    pages_iter
```
Comment (gshep) [inline core/src/memory.rs:189]: Frankly speaking it looks like there should be guarantee that `pages_iter` is sorted

File: runtime-interface/src/lib.rs [hunk 4]
```diff
@@ -171,42 +171,37 @@ pub trait GearRI {
...
-        let mut prev_page = 0u32;
-        let mut size_in_pages = 0u32;
-        let mut lazy_pages = gear_lazy_pages::get_lazy_pages_numbers();
+        let lazy_pages = gear_lazy_pages::get_lazy_pages_numbers();
+        if lazy_pages.is_empty() {
+            return Ok(());
+        }
 
         if let Some(page) = lazy_pages.last() {
             // This case is impossible in real live, so just returns err and does nothing.
```
Comment (gshep) [inline runtime-interface/src/lib.rs:180]: I know this is not your changes nevertheless maybe edit the text to `... impossible in reality`?

**Review by gshep — APPROVED** [review]

File: runtime-interface/src/lib.rs (hunk 1 above)
Comment (gshep) [reply runtime-interface/src/lib.rs:189]: we definitely need a test(s) for comfortable refactoring =)

File: core/src/memory.rs (hunk 2 above)
Comment (shamilsan) [reply core/src/memory.rs:192]: It is possible if the input value is not aligned (see the example that results in error).

File: core/src/memory.rs (hunk 3 above)
Comment (shamilsan) [reply core/src/memory.rs:189]: Exactly. I've not changed the function contract.

File: runtime-interface/src/lib.rs (hunk 4 above)
Comment (shamilsan) [reply runtime-interface/src/lib.rs:180]: I'd remove this part at all 😄 But I agree to rename at least.

File: runtime-interface/src/lib.rs (hunk 1 above)
Comment (shamilsan) [reply runtime-interface/src/lib.rs:189]: I don't know how to efficiently test mprotect. Need assistance from @grishasobol

File: runtime-interface/src/lib.rs [hunk 5]
```diff
@@ -171,42 +171,37 @@ pub trait GearRI {
...
 
         // Collects continuous intervals of memory from lazy pages to protect them.
-        for page in lazy_pages {
-            if prev_page + 1 == page {
-                size_in_pages += 1;
+        let mut start = lazy_pages[0] as usize;
+        let mut count = 1;
+        for page in lazy_pages.into_iter().skip(1) {
+            if start + count == page as usize {
+                count += 1;
```
Comment (grishasobol) [inline runtime-interface/src/lib.rs:197]: Do we wanna to hold all arith cases? For example here can be count overflow

**Review by grishasobol — APPROVED** [review]

File: runtime-interface/src/lib.rs (hunk 5 above)
Comment (shamilsan) [reply runtime-interface/src/lib.rs:197]: Yes, if we remove the last element check against `u32::MAX`, `count` can exceed `u32::MAX`for `[0..u32::MAX]` input array. Will fix it.

File: core/src/gas.rs [hunk 6]
```diff
@@ -74,6 +74,7 @@ impl GasCounter {
         }
 
         self.left -= amount;
+        // `burned` can't overflow as `amount` can't be greater than u64::MAX
```
[the remark this answers is not in the document]
Comment (shamilsan) [reply core/src/gas.rs]: I've added a comment.

File: runtime-interface/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (shamilsan) [reply runtime-interface/src/lib.rs]: Removed.

File: runtime-interface/src/lib.rs [hunk 7]
```diff
@@ -171,42 +171,37 @@ pub trait GearRI {
...
+            let addr = wasm_mem_addr + (start * PageNumber::size()) as u64;
+            let size = count * PageNumber::size();
+            sys_mprotect_interval(addr, size, !protect, !protect, false)
+        };
 
         // Collects continuous intervals of memory from lazy pages to protect them.
-        for page in lazy_pages {
-            if prev_page + 1 == page {
-                size_in_pages += 1;
+        let mut start = lazy_pages[0] as usize;
```
[the remark this answers is not in the document]
Comment (shamilsan) [reply runtime-interface/src/lib.rs]: Added a comment.

**Pull request merged**